### ISSUE TITLE
GitBuildVersion not being set when run on self-hosted agent

### ISSUE THREAD

**Issue by rogerstewart** [dotnet/Nerdbank.GitVersioning#531]
Today I updated the projects in my solution to NBGV 3.3.37 nuget package from 3.0.48. This was the only change made.

My deployment builds, run on a **self-hosted agent**, no longer show the 'pretty' version in my Azure pipeline history and has reverted to the 20201030.1 format.

More importantly some of my artifacts rely on the **GitBuildVersion** variable and no longer set. My artifact names now look like **product-x86-$(GitBuildVersion).zip** instead of **product-x86-1.0.22.zip**.
The log no longer outputs lines like the following:
`2020-10-27T17:34:14.0357436Z ##[section]Async Command Start: Update Build Number`
`2020-10-27T17:34:14.0358030Z Update build number to 1.0.22 for build 9534`
`2020-10-27T17:34:14.0358433Z ##[section]Async Command End: Update Build Number`

My CI builds, that are run on Azure cloud hosted agents, get their build version updated and artifact names are as expected. Its just when I run the deploy script on the self-hosted agent is when I see the issue.

I checked the **doc/cloudbuild.md** documentation and did not see any breaking changes from older versions. I build with .NET Framework and do not want to add tasks to install dotnet core just to run the set variables script.

**Comment by AArnott** [comment]
Very peculiar. On your self-hosted agent, is that agent being driven by Azure Pipelines then? 
How are you setting the cloud build variables? Are you using the `nbgv cloud` command or the nerdbank.gitversioning nuget package with `version.json` configured to set cloud variables?

Can you perhaps link to the source code or pipeline logs?

**Comment by rogerstewart** [comment]
> On your self-hosted agent, is that agent being driven by Azure Pipelines then?

Yes The only difference between my release.yaml and my ci.yaml is the pool vmImage setting.

> Are you using the nbgv cloud command or the nerdbank.gitversioning nuget package with version.json configured to set cloud variables?

I am using the nerdbank.gitversioning nuget package with a single root version.json configured to set cloud variables.

I cannot link to the source or logs due to this is a product repo. I could provide snippets of the logs or NBGV related project settings if needed. I would need guidance though as to what to snip from the logs.

My solution consists of 3 MSVC C++ projects, a WiX setup project, and a WiX bootstrapper project. 
- The bootstrapper project contains a post build step that uses the GitBuildVersion to rename the installer exe successfully. So I know the build can see and use the variable. 
- However, once the yaml VSBuild@1 task completes a subsequent ArchiveFiles@2 task fails to get the GitBuildVersion to name the zip archive.
- I don't know if this matters or not, but I use Azure yaml templates. The build task is in one template file and the archive in another.

**Comment by AArnott** [comment]
I use YAML for Azure Pipelines all the time. But I don't often use the cloud build variables. I do often set the cloud build number though, and haven't noticed any regression. 
Since you say that "Update build number to 1.0.22 for build 9534" doesn't show up in your log any more _and_ all your cloud build variables aren't set, it sounds like Nerdbank.GitVersioning isn't successfully setting it any more. Are there _errors_ or warnings in your log anywhere around a failed attempt? Or perhaps it isn't set at all.

If you don't see any errors/warnings, I suspect NB.GV doesn't recognize that it's running in Azure Pipelines so it doesn't issue the console commands to set the cloud build number and variables.
I suggest you add a `SET` command (or similar) to dump all environment variables to the start of your pipelines. Compare the Hosted agent and private agent pipeline results. NB.GV specifically looks for `SYSTEM_TEAMPROJECTID` to detect Azure Pipelines in action:

https://github.com/dotnet/Nerdbank.GitVersioning/blob/d9c51b8d5c654f722fafdf9e9f4c721e401fd7ed/src/NerdBank.GitVersioning/CloudBuildServices/VisualStudioTeamServices.cs#L24

**Comment by rogerstewart** [comment]
Both hosted and self-hosted logs contain `SYSTEM_TEAMPROJECTID=<some-guid>`.

After building the solution, only the hosted log contains:
`GITASSEMBLYINFORMATIONALVERSION=1.0.40+3a2409c2ac`
`GITBUILDVERSION=1.0.40.9274`
`GITBUILDVERSIONSIMPLE=1.0.40`

**Comment by AArnott** [comment]
Very strange.
Can you set the `SYSTEM_TEAMPROJECTID` env var on your own box in an command window and build one of your projects there? Then look at your console and/or build log file to see if you can find any `##vso` console lines written out?

**Comment by rogerstewart** [comment]
I see `##vso[task.logdetail` and `##vso[task.logissue` entries. I do not see any `##vso[build.updatebuildnumber]` or `##vso[task.setvariable` entries.

**Comment by rogerstewart** [comment]
Following your steps of adding `SYSTEM_TEAMPROJECTID` as an environment var and building my solution from a VS command prompt, have found how to reproduce this on my dev and deploy machines. 

To see it working as expected I built my solution with a usual command line:
`msbuild my.sln /p:platform="x86" /p:configuration="Release"`

To reproduce the issue, I enable parallel builds with the /m switch:
`msbuild my.sln /m /p:platform="x86" /p:configuration="Release"`

The yaml `VSBuilds` task `maximumCpuCount: true` adds the `/m` switch.

**Comment by AArnott** [comment]
Wow. The `/m` switch causes the project to not have `SYSTEM_TEAMPROJECTID` defined? What version of msbuild are you running? That's a bug if the msbuild.exe worker processes do not reapply the environment variables from the original process.

**Comment by rogerstewart** [comment]
`SYSTEM_TEAMPROJECTID` is always defined. It is present in environment dumps before and after `VSBuild` task. 

The /m switch causes NBGV 3.3.37 to not add its cloud variables via inserting `##vso[build.updatebuildnumber]` or `##vso[task.setvariable` entries. Without the /m switch they get added as expected.

NBGV 3.0.48 works as expected with or without the /m switch.

**Comment by AArnott** [comment]
> The /m switch causes NBGV 3.3.37 to not add its cloud variables

NBGV has no idea whether the /m switch is specified.  It only reads files and environment variables.

> SYSTEM_TEAMPROJECTID is always defined. It is present in environment dumps before and after VSBuild task.

What I mean is we need to dump the environment variables in each msbuild.exe process kicked off by `/m`. I get that it's set before and after the "VSBuild" task. But what about in the process when NBGV is running?
Can you add this target to your project:

```xml
  <Target Name="DumpEnv" BeforeTargets="GetBuildVersion">
    <Exec Command="SET" StandardOutputImportance="high" />
  </Target>
```

Then run it locally with and without `/m` (with `SYSTEM_TEAMPROJECTID` set before-hand) and share whether you see the env var in the log?
Also try running this in AzP to see what the result it.

**Comment by AArnott** [comment]
I can't explain how this behavior changed from 3.0.48 to 3.3.37 yet. But as it's easier to add logging to msbuild than to a compiled dll like nerdbank.gitversioning.dll, it seems reasonable to start here.

**Comment by rogerstewart** [comment]
It appears that 'SYSTEM_TEAMPROJECTID' is set everytime 'DumpEnv' is called when run locally or through Azure Pipelines with or without /m.

I have attached a azure pipeline raw log of a build with /m containing the DumpEnv info.
[dump_scrub.txt](https://github.com/dotnet/Nerdbank.GitVersioning/files/5476364/dump_scrub.txt)

**Comment by AArnott** [comment]
I cannot imagine why this is broken, and I can't repro it. Here is what I tried:

```ps1
dotnet new classlib
git init
copy \some\.gitignore
copy \some\.gitattributes
git add .
git commit -m "initial commit"
nbgv install
git commit -m "add nb.gv"
$env:system_teamprojectid='123'
dotnet build /m
```

Build output

```
Microsoft (R) Build Engine version 16.8.0-preview-20475-05+aed5e7ed0 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored D:\temp\nbgvtest\nbgvtest.csproj (in 88 ms).
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
##vso[task.setvariable variable=GitAssemblyInformationalVersion;]1.0.1-beta+1ffe3abb59
##vso[task.setvariable variable=GitBuildVersion;]1.0.1.65055
##vso[task.setvariable variable=GitBuildVersionSimple;]1.0.1
##vso[build.updatebuildnumber]1.0.1-beta
  nbgvtest -> D:\temp\nbgvtest\bin\Debug\net5.0\nbgvtest.dll

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.17
```

So this looks good to me, since I see the `##vso` logging commands.

Since you can't share the repo or .binlog, can you provide a minimal repro so I can see it locally?

**Comment by rogerstewart** [comment]
I don't know what a .binlog is nor can find one in my repo dir structure. So, I will try to create a repro.

**Comment by AArnott** [comment]
Thanks. 

FWIW You can learn about .binlog here: https://www.msbuildlog.com/
But it discloses just about everything in the build and isn't in a format friendly to you redacting bits you're sensitive about. So I don't think you should share it with me.

**Comment by rogerstewart** [comment]
I am still trying to get a repo together.

I did use the `/bl` switch and msbuildlog.com tool to look at a build where the `##vso` output was missing. The SetCloudVariable related tasks were called as expected with the correct output values. Strange. Too bad the log doesn't show where the tasks accessed stdout/Console and if issues occurred doing so.

**Comment by rogerstewart** [comment]
I was able determine and fix the issue on my side. Unfortunately, I was unable to reproduce in a slimmed down repo.

**Comment by julealgon** [comment]
@rogerstewart what was the problem on your side? I'm seeing intermittent behavior where some of my Azure DevOps builds are missing the version update section which is very similar to your original problem here.

I've not been able to identify any differences between a successful build and a problematic one however... nothing jumps out to me.

**Comment by rogerstewart** [comment]
@julealgon two things helped my situation.

1. I had to turn off parallel builds in my Azure CI/CD builds by setting the maxcpucount switch to false. In the yaml VSBuild task it is maximumCpuCount.
2. More specific to my case, I also had Project Dependencies set for some projects that no longer needed them.

Unfortunately to this day, when running CI/CD jobs in Azure DevOps, I have not been able to build the projects in my solutions in parallel and receive output with the expected version information. With that said, though, I am running older NGV versions and have not had the time to try upgrading to see if my issues would be resolved.

**Comment by AArnott** [comment]
FWIW I use NB.GV in all my repos, with parallel builds turned on. I haven't had any problems. We'd need to look at your diagnostic logs to find out what's going wrong in your case.
